####-Pulse-Channels-IR-and-Compiler.md

```diff
@@ -0,0 +1,277 @@
+- IR Instruction - A pulse program instruction. Typically, it will be initialized without concrete timing, and will be scheduled during the compiler operation.
+
+Two instructions types are used:
+- `GenericInstruction` - for play, delay, set\shift frequency\phase. Assignment of `LogicalElement` will be optional for set\shift frequency\phase and will determine whether or not the instruction will be
+broadcaseted to all mixed frames associated with the frame or not.
+- `AcquireInstruction` - for acquire instructions (due to the different elements involved).
+
+```
+class PulseIR:
+    def __init__(self):
+        self.instructions # List of instructions or nested PulseIR objects
+        self.alignment  
+
+class GenericInstruction:
+    def __init__(
+        self,
+        instruction_type: str, 
+        duration: int, 
+        logical_element: Optional[LogicalElement] = None, 
+        frame: Optional[Frame] = None, 
+        **operands,
+    ):  
+        self.t0 = None
+        self.instruction = instruction_type  # opcode
+        self.duration = duration
+        self.logical_element = logical_element
+        self.frame = frame
+        self.operands = operands
+```
+
+### Compiler
+The compiler will be built with a shared design of Qiskit's pass manager (based on PR #10474), and will thus be easily extendable to other tasks. The pass based approach allows for simple modification of the lowering process for vendors or Qiskit team. 
+A typical compiler call will look like:
+```
+payload = compile(my_pulse_prog, backend, target="pulse_qobj")
+```
+where the backend is provided to accomodate the compiled program to the backend constraints and mappings. An optional argument will be a mapping between logical and physical qubits, as under the new model all indices will be assumed logical. If no mapping is provided, the trivial one will be used.
+
+The first step of every compiler run will be to initialize the IR of the pulse program. The IR will be initialized with no concrete timing, and the scheduling will be carried out next.
+
+Next we highlight some of the tasks handled by the compiler.
+
+#### Scheduling
+- Canonicalization (transform pass) - If legacy `Channel`s were used, they will be mapped to `MixedFrame`s using the backend mapping. This is the only scenario where backend mapping will be needed for this stage.
+- Identify frames and mixed frames (characterization pass) - `FrameInstruction`s will need to be broadcasted to every `MixedFrame` associated with that `Frame`. Therefore, we need identify every `MixedFrame`, `Frame` and the relations between them.
+- Schedule each block -
+  - Starting from the lowest blocks, each block will be scheduled on its own.
+  - Set\shift frequency\phase instructions will be broadcasted when necessary, and the additional mixed frames will be added to the block.
```

Comment:
We can add analysis pass that lists every mixed frame before doing broadcasting the instruction. I think this is more efficient than doing it on the fly.

Reply:
I agree, and that's why I listed it a few steps earlier :) 
But there were some inaccuracies there which I had to be corrected.